### ISSUE TITLE
Instruction how build for ios and android

### ISSUE THREAD

**Issue by sergeymild** [libcpr/cpr#715]
Hello, maybe i am stupid), but i don't understand how build this library for ios and android, it would be nice to have some kind of instructions. Thanks.

**Comment by COM8** [comment]
I've never done this for those platforms either. Sooo I need a bit more context to help you with that e.g.:
* What is your build system?
* What OS?
* Any errors?
* What did you try?

**Comment by sergeymild** [comment]
My build system: Cmake
What OS: Mac
Any errors: i tried build for ios and receive this error: `ld: library not found for -lsocket`
What did you try:
git clone https://github.com/libcpr/cpr
after download ios.toolchain.cmake `[os.toolchain.cmake](https://github.com/leetal/ios-cmake/blob/master/ios.toolchain.cmake)`
and put to a folder where i dowloaded cpr
after i did everything by doc

cd cpr
mkdir build
cd build
cmake .. -G Xcode -DCMAKE_TOOLCHAIN_FILE=../ios.toolchain.cmake -DPLATFORM=SIMULATOR64
cmake --build . --config Release

and receive this error:


```
Ld /Users/sergeigolishnikov/develpment/cpr/build/lib/Release/libcurl.dylib normal (in target 'libcurl' from project 'cpr')
    cd /Users/sergeigolishnikov/develpment/cpr
    /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang -target x86_64-apple-ios11.0-simulator -dynamiclib -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator15.2.sdk -L/Users/sergeigolishnikov/develpment/cpr/build/lib/Release -F/Users/sergeigolishnikov/develpment/cpr/build/lib/Release -filelist /Users/sergeigolishnikov/develpment/cpr/build/_deps/curl-build/lib/cpr.build/Release-iphonesimulator/libcurl.build/Objects-normal/x86_64/libcurl.LinkFileList -install_name @rpath/libcurl.dylib -dead_strip -Xlinker -objc_abi_version -Xlinker 2 -rpath @executable_path/Frameworks -rpath @loader_path/Frameworks -dynamiclib -Wl,-headerpad_max_install_names -lsocket -framework Security -framework CoreFoundation /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator15.2.sdk/usr/lib/libz.tbd -Xlinker -sectcreate -Xlinker __TEXT -Xlinker __entitlements -Xlinker /Users/sergeigolishnikov/develpment/cpr/build/_deps/curl-build/lib/cpr.build/Release-iphonesimulator/libcurl.build/libcurl.dylib-Simulated.xcent -Xlinker -no_adhoc_codesign -Xlinker -dependency_info -Xlinker /Users/sergeigolishnikov/develpment/cpr/build/_deps/curl-build/lib/cpr.build/Release-iphonesimulator/libcurl.build/Objects-normal/x86_64/libcurl_dependency_info.dat -o /Users/sergeigolishnikov/develpment/cpr/build/lib/Release/libcurl.dylib
ld: library not found for -lsocket
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

**Comment by COM8** [comment]
Please try `-DENABLE_STRICT_TRY_COMPILE=ON` when configuring.

References:
* https://gitlab.kitware.com/cmake/cmake/-/issues/21971
* https://github.com/curl/curl/issues/6789

**Comment by sergeymild** [comment]
yes it's works, it generated `libcpr.1.8.0.dylib`, but i need .a or .so, maybe you know what i need to do to generate .a or .so?

**Comment by sergeymild** [comment]
the same was build for android as well

**Comment by COM8** [comment]
Well I don't think this is a cpr issue then and more a general build system issue where you do not target the correct system.

**Comment by sergeymild** [comment]
i specified it here cmake .. -G Xcode -DCMAKE_TOOLCHAIN_FILE=../ios.toolchain.cmake -DPLATFORM=SIMULATOR64
-DPLATFORM=SIMULATOR64 - where is simulator x86_86

**Comment by ospfranco** [comment]
@sergeymild did you manage to compile the lib for iOS?

**Comment by ospfranco** [comment]
So following similar steps to @sergeymild I have managed to compile (I had to manually fix an implicit conversion) and cmake has produced a xcodeproj. As far as I understood it I'm supposed to add this project as a dependency, which I have, but I still cannot import the header (`#import <cpr/cpr.h>`) into my code. I know this is is very specific to iOS/Xcode and the authors might not know the answer. But more specific instructions would be greatly appreciated.

This other [issue](https://github.com/libcpr/cpr/issues/113) is related, he also seems to have used cmake but for him it did generate an .a binary.

Would really like to get some support here :)

**Comment by ospfranco** [comment]
I have created a test repo if someone is interested in taking a look:

https://github.com/ospfranco/libcprtest

As I said, I followed the same steps sergey posted to generate the .dylib, but then copied the cpr repo inside the main repo so I can include the headers (I had to manually modify the files so that the imports use quotes instead of angle brackets). So now the app compiles and it launches, but it immediately crashes. Probably because the linked libraries are not correct?

**Comment by COM8** [comment]
Although I never developed for Android or iOS this is probably something like it can't find the libs.

Let me know when you get it up and running, so we can include such a sample in our samples collection here: https://github.com/libcpr

**Comment by ospfranco** [comment]
I've been doing a lot of reading on CMake to try to really understand how you are supposed to build iOS libraries. Here are some of my findings.

- Using the `-g Xcode` (Xcode generator) not only targets apple systems, it will actually generate files you can directly import into Xcode, in this case an `.xcodeproj`, it does not necessarily output static or dynamic libraries
- libcpr cmakelists includes a `BUILD_SHARED_LIBS` flag (read CMake option, can be set to OFF) that will output libcpr, libcurl and libz as static .a files: `cmake .. -G Xcode -DCMAKE_TOOLCHAIN_FILE=../ios.toolchain.cmake -DPLATFORM=OS64COMBINED -DENABLE_STRICT_TRY_COMPILE=ON -DDEPLOYMENT_TARGET=13.0 -DBUILD_SHARED_LIBS=OFF`
- It seems to me trying to import just the `.a` files defeats the purpose of the Xcode generator, one is rather supposed to add the generated `.xcodeproj` as a subproject.
- I have, however, tried both methods (copying static `.a` files and importing the generated `.xcodeproj`) and it *seems to me* that both are fine (Xcode did complain when I built the `.a` files for the wrong target).
- I am however stuck with header includes, for some reason the generated `.xcodeproj` does not include headers into itself. No matter how I try (adding path to `User Header Search Paths`, manually copying them) Xcode is not happy when I do `#import "cpr/cpr.h"`. In the case of manual copy/pointing-to the cpr/include I've managed to get the cpr headers recognized, but then the libcurl headers are missing.

**Comment by ospfranco** [comment]
I got it working! Andre answered my [Stack Overflow question](https://stackoverflow.com/questions/73216268/how-to-include-xcode-sub-project-headers) and gave me the exact values I needed to fetch into Xcode to get it to detect the headers.

Now I'm facing another issue, it seems SSL requires a `.pem` file to be detected in the OS for https (TLS) requests to be made. I saw some other issues in the repo that point to possible solutions, but I don't quite understand what is the problem.

<img width="660" alt="Screenshot 000824" src="https://user-images.githubusercontent.com/1634213/182685328-4a889f7d-670f-4024-a189-5d78b5e45e03.png">

Any idea how to fix this?

**Comment by ospfranco** [comment]
It seems on iOS the root CA is not accessible (due to iOS apps being sandboxed and what not). [This StackOverflow answer](https://stackoverflow.com/questions/42432473/programmatically-read-root-ca-certificates-in-ios) seems to suggest the usual approach is to download the apple root cert in the app bundle and reading it from there... which seems complex to me. Maybe there is no other way. Any suggestions on how to set the ca cert on libcurl?

**Comment by ospfranco** [comment]
It seems there are a couple of solutions:
1. Bundle your own certificate and pass it to libcurl on runtime (I would not want to do this due to app review)
2. Build libcurl with the `--with-secure-transport` flag, which [will use Apple's Secure Transport](https://curl.se/docs/install.html) (search for the iOS/macOS part) and it will automagically use the certificates used on the device keychain

I'm trying to make 2 work but not sure how to compile libcurl with that flag. The [cmakelists](https://github.com/libcpr/cpr/blob/master/CMakeLists.txt#L250) already appears to be setting some option. Any ideas?

**Comment by ospfranco** [comment]
Just checked the CMake output one more time and Secure Transport seems to be used... weird that iOS fails to load the cert then...

<img width="576" alt="Screenshot 000825" src="https://user-images.githubusercontent.com/1634213/182763533-6d25f13b-02e2-48bc-a7cc-284c9d6accb2.png">

**Comment by ospfranco** [comment]
[This mailing list](https://curl.se/mail/lib-2017-11/0070.html) seems to suggest some flags (CURLOPT_CAINFO to NULL) need to be turned off when not using the shared curl library (also, any idea how to do that?)

I might also try not using a static library build?

**Comment by ospfranco** [comment]
Scratch that last comment it worked! W00p w00p!

![Screenshot 000827](https://user-images.githubusercontent.com/1634213/182772982-ac21dc05-767d-44b0-99a4-9ba8338f969d.png)

**Comment by sergeymild** [comment]
@ospfranco it's awesome, could you please also write some kind of instruction how to build it on mac for ios? You will help me a lot) because now, to build my react-native-cpr i need to copy all cpr library and distribute it together.

**Comment by ospfranco** [comment]
The instructions are all up there ⬆️ I will try to write something more clean later :)

**Comment by ospfranco** [comment]
I updated my submodule to the latest version (since merging #801) but now when I compile the project the generated .xcodeproject contains a `libcurl-d.a` which is not anywhere in the generated files:

<img width="1234" alt="Screenshot 000860" src="https://user-images.githubusercontent.com/1634213/184103909-a0f0d492-2c38-45ba-b0ca-cb2ec30c2a85.png">

<img width="924" alt="Screenshot 000861" src="https://user-images.githubusercontent.com/1634213/184103985-f0186f39-3093-4562-ae9c-56d4768101d1.png">

Any idea if anything changed in the last days that might change the outputted libraries?

**Comment by ospfranco** [comment]
CMake is a funny beast... it seems it might read the parent xcodeproj to determine which build to generate? so libcurl-d.a seems to be a debug build... but when I link it in iOS I don't want debug builds to ship on production...

<img width="1677" alt="Screenshot 000863" src="https://user-images.githubusercontent.com/1634213/184118879-c01e86b7-f334-4afc-9a98-d3eca7dc613d.png">

**Comment by COM8** [comment]
It would be awesome if someone could create a PR for the https://github.com/libcpr/docs repo with a more detailed summary/run through description for this.

**Comment by ospfranco** [comment]
I will do, but did you see my last comment? is it correct that a debug `.a.` is generated? or how can I avoid this?

**Comment by COM8** [comment]
In my eyes this is not correct. Perhaps there is something wrong in curl. Did you already try to build libcurl without cpr?

**Comment by ospfranco** [comment]
Sorry, I was busy with other stuff, taking a look into this right now.

No, libcurl is being compiled by the base CMakeLists of cpr, I did not change anything. I went over the file but did not find any obvious flag to compile libcurl in release mode, the thing is both versions are generated, so probably it is the toolchain file that needs to generate the xcodeproj pointing to the correct configuration.

Regarding the compilation, I do not have time for something more detailed, but now that my changes are merged here is the basic rundown:

- Add `libcpr` as a git submodule
- Add the [ios.toolchain.cmake](https://github.com/leetal/ios-cmake/blob/master/ios.toolchain.cmake) file in the root directory
- `cd cpr && mkdir build && cd build && cmake .. -G Xcode -DCMAKE_TOOLCHAIN_FILE=../../ios.toolchain.cmake -DPLATFORM=OS64 -DDEPLOYMENT_TARGET=13.0 -DBUILD_SHARED_LIBS=OFF -DCPR_SKIP_CA_BUNDLE_SEARCH=ON && cmake --build . --config Release`
- On Xcode modify the `Header Search Paths` on the Build Settings, add a new entry with `"$(SRCROOT)/../cpr"
- On the project/file explorer, drag and drop the generated `.xcodeproj` into the project (I placed into a "Libraries" directory/group)
- Including and compiling `<cpr/cpr.h>` should work

**Comment by COM8** [comment]
Thanks for taking the time to note down your steps! I will try to move them inside the docs.

**Comment by cpsauer** [comment]
@ospfranco, just to clarify, DBUILD_SHARED_LIBS=FALSE currently needs to be specified for things to work with sandboxed Apple OSs, right?

Would it makes sense to automatically default it to OFF on Apple Platforms?

[I'm working on writing Bazel rules to build CPR across platforms. @COM8, lmk if you'd be interested in them as a contribution. Benefits would be easy CPR usage for more users, especially bigcos that might contribute.]

**Comment by ospfranco** [comment]
TBH no idea if SHARED_LIBS=OFF is necessary. I wanted a static library because I thought it would be easier that way to include it since I didn't know if the produced dylibs might cause compatibility issues. I don't think it would make a difference, since internally the compiled library still uses Apple's network layer, so they can still track/block as much as they want.

**Comment by COM8** [comment]
@cpsauer since you are working on Bazel rules anyway please let me know when you are done, so I can have a look at the required changes. The issue here is, this would require maintenance later on and I currently don't have the knowledge required about the Bazel build system.

**Comment by cpsauer** [comment]
@COM8, will do and totally understand. I suspect it'll be much simpler/shorter than the equivalent cmake.

**Comment by cpsauer** [comment]
@ospfranco, sorry, I messed up. 
I meant to say -DCPR_SKIP_CA_BUNDLE_SEARCH--proposing that maybe it should default to on for Apple platforms if its needed?
Have edited the above to make sense. Again, my apologies.

**Comment by COM8** [comment]
Perhaps one can even provide a small sample XCode project like we have done it for [CMake](https://github.com/libcpr/example-cmake-fetch-content) and [Conan](https://github.com/libcpr/example-cmake-conan).

**Comment by ospfranco** [comment]
no, on macOS the bundle can be accessed without problems, only on iOS it causes problems because the Sandboxing. I already provided [the sample project](https://github.com/ospfranco/libcprtest) where I ran the integration tests in case that is useful to you.

**Comment by cpsauer** [comment]
[Had to sleep. Pretty cool that we're on opposite sides of the world and still working together!]

@ospfranco, I'm trying to say something a bit different. Sorry I haven't communicated very well so far. To restart from scratch: 
1. From the above discussion it looks like -DCPR_SKIP_CA_BUNDLE_SEARCH=ON is needed on iOS.
2. That made me think: Perhaps the cmake file should therefore automatically set CPR_SKIP_CA_BUNDLE_SEARCH=ON for iOS, so things work automatically for people. I figured I'd do the equivalent for Bazel.
3. I'd guess the same is true for the other (very similar) sandboxed Apple OSs (watchOS, tvOS). 
4. And since macOS has a sandboxed mode for the App Store, perhaps we should configure it the same way, to use secure-transport--if it works. That said, since libcurl is part of the macOS API, maybe we should just use that as the working default.

[@COM8, I'm assuming that the Xcode project suggestion is for ospfranco, but if not, the Bazel stuff will be cross-platform and cross-editor.]

**Comment by ospfranco** [comment]
I know what you meant, just pointing the differences, feel free to do the changes, but without the toolchain file (which is huge) I don't know how to generate pure static files, so it seems to me like more changes are needed than just one flag :)

**Comment by cpsauer** [comment]
For any Bazel users who come across this: 
rules supporting iOS and Android: https://github.com/hedronvision/bazel-make-cc-https-easy

**Comment by tpecholt** [comment]
I believe this should be documented somewhere.

As for the android these steps have sort of worked for me:

### OpenSSL doesn't come with CMakeLists. Build it manually with your android-ndk:
```
export ANDROID_NDK_ROOT=~/android-ndk-r27d
PATH=$ANDROID_NDK_ROOT/toolchains/llvm/prebuilt/linux-x86_64/bin:$ANDROID_NDK_ROOT/toolchains/arm-linux-androideabi-4.9/prebuilt/linux-x86_64/bin:$PATH
cd openssl
./Configure android-arm64 -D__ANDROID_API__=21
make
```
then copy .so files to android/app/src/main/jniLibs/arm64-v8a folder and include/openssl into your android/app/src/main/cpp/openssl

### Your android native lib CMAKE:
```
# find_package won't work set targets and output variables manually:
add_library(OpenSSL::SSL SHARED IMPORTED)
set_target_properties(OpenSSL::SSL PROPERTIES IMPORTED_LOCATION
        ${CMAKE_CURRENT_SOURCE_DIR}/../jniLibs/${ANDROID_ABI}/libssl.so)
set_target_properties(OpenSSL::SSL PROPERTIES INTERFACE_INCLUDE_DIRECTORIES
        ${CMAKE_CURRENT_SOURCE_DIR})

add_library(OpenSSL::Crypto SHARED IMPORTED)
set_target_properties(OpenSSL::Crypto PROPERTIES IMPORTED_LOCATION
        ${CMAKE_CURRENT_SOURCE_DIR}/../jniLibs/${ANDROID_ABI}/libcrypto.so)

# CPR library

set(CPR_CURL_USE_LIBPSL OFF)
# cpr uses OpenSSL targets these are just for checking OpenSSL was found:
set(OPENSSL_INCLUDE_DIR .)
set(OPENSSL_CRYPTO_LIBRARY libcrypto.so)
add_subdirectory(../../../../../3rdparty/cpr cpr)

then target_link_libraries with cpr::cpr
```

But it's not finished. Currently I am getting a compiler error:
```
curl-src/lib/vtls/openssl.c:413:11: error: incomplete definition of type 'struct asn1_string_st'
  413 |     if(num->type == V_ASN1_NEG_INTEGER)
     |        ~~~^
C:/work/game-recorder/android/app/src/main/cpp/openssl/types.h:58:16: note: forward declaration of 'struct asn1_string_st'
   58 | typedef struct asn1_string_st ASN1_INTEGER;
      |                ^
```
